tools/smixlate.1.in
.\"
.\" $Id: smixlate.1.in 1676 2004-08-10 10:58:12Z strauss $
.\"
.TH smixlate 1  "June 18, 2006" "IUB" "SMI Tools"
.SH NAME
smixlate \- translate SMI/SPPI identifiers
.SH SYNOPSIS
.B smixlate
[
.B "-Vhm"
] [
.BI "-c " file
] [
.BI "-p " module
] [
.BI "-l " level
]
.I "module(s)"
.SH DESCRIPTION

The \fBsmixlate\fP program is used to translate identifiers and
especially OIDs into a more human readable format.
.SH OPTIONS
.TP
\fB-V, --version\fP
Show the smixlate version and exit.
.TP
\fB-h, --help\fP
Show a help text and exit.
.TP
\fB-r, --recursive\fP
Report errors and warnings also for recursively imported modules. 
.TP
\fB-c \fIfile\fB, --config=\fIfile\fP
Read \fIfile\fP instead of any other (global and user)
configuration file.
.TP
\fB-p \fImodule\fB, --preload=\fImodule\fP
Preload the module \fImodule\fP before reading the main
module(s). This may be helpful if an incomplete main module misses to
import some definitions.
.TP
\fB-l \fIlevel\fB, --level=\fIlevel\fP
Report errors and warnings up to the given severity \fIlevel\fP. See
the smilint(1) manual page for a description of the error levels. The
default error level is 3.
.TP
.I module(s)
These are the modules to be loaded for the subsequent translation. If
a module argument represents a path name (identified by containing at
least one dot or slash character), this is assumed to be the exact
file to read. Otherwise, if a module is identified by its plain module
name, it is searched according to libsmi internal rules. See
\fBsmi_config(3)\fP for more details.
.SH "EXAMPLE"
This example translates numeric OIDs in the input text into a more
human readable format.
.nf

  $ echo "what is this oid? 1.3.6.1.2.1.2.2.1.3" | \    
    ./smixlate -l 0 /usr/local/share/mibs/ietf/*
  what is this oid? ifType
  $

.fi
.SH "SEE ALSO"
The 
.BR libsmi (3)
project is documented at
.BR "http://www.ibr.cs.tu-bs.de/projects/libsmi/" "."
.SH "AUTHORS"
(C) 2006-2006 J. Schoenwaelder, International University Bremen, Germany
.br
and contributions by many other people.
.br


ChangeLog
2005-06-12  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/ISCSI-MIB: added (RFC 4544).
	* mibs/ietf/IPS-AUTH-MIB: added (RFC 4545).

2005-05-30  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/SNMP-USM-AES-MIB: updated (RFC 3826).

2005-05-29  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/BGP4-MIB: updated (RFC 4273).
	* mibs/ietf/IP-FORWARD-MIB: updated (RFC 4292).
	* mibs/ietf/IP-MIB: updated (RFC 4293).
	* mibs/ietf/MOBILEIPV6-MIB: added (RFC 4295).
	* mibs/ietf/RSTP-MIB: added (RFC 4318).
	* mibs/ietf/DOCS-IETF-QOS-MIB: added (RFC 4323).
	* mibs/ietf/LMP-MIB: added (RFC 4327).
	* mibs/ietf/{P-BRIDGE-MIB,Q-BRIDGE-MIB}: updated (RFC 4363).
	* mibs/ietf/{MPLS-LC-ATM-STD-MIB,MPLS-LC-FR-STD-MIB):
	  added (RFC 4368).
	* mibs/ietf/IFCP-MGMT-MIB: added (RFC 4369).
	* mibs/ietf/MPLS-L3VPN-STD-MIB: added (RFC 4382).
	* mibs/ietf/FCIP-MGMT-MIB: added (RFC 4404).
	* mibs/ietf/T11-FC-NAME-SERVER-MIB: added (RFC 4438).
	* mibs/ietf/{T11-TC-MIB,T11-FC-FABRIC-ADDR-MGR-MIB}:
	  added (RFC 4439).
	* mibs/ietf/ISIS-MIB: added (RFC 4444).
	* mibs/ietf/SCSI-MIB: added (RFC 4455).
	* mibs/ietf/{AGGREGATE-MIB,TIME-AGGREGATE-MIB}: added (RFC 4498).
	* mibs/ietf/RMON2-MIB: updated (RFC 4502).

	* libsmi-0.4.5.tar.gz: released version 0.4.5.
	There was a bug that breaks compliation of 0.4.4, so here
	is a new release.

2006-05-16  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* libsmi-0.4.4.tar.gz: released version 0.4.4.
	This is more or less just a snapshot to give package
	maintainers a fresh and handy release number.

2005-12-18  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/HDSL2-SHDSL-LINE-MIB: updated (RFC 4319).

2005-12-16  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/{ENTITY-STATE-MIB,ENTITY-STATE-TC-MIB}:
	  added (RFC 4268).

2005-11-25  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/TE-LINK-STD-MIB: added (RFC 4220).

	* mibs/ietf/VPN-TC-STD-MIB: added (RFC 4265).

2005-10-10  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/DOCS-IETF-BPI2-MIB: added (RFC 4131).

	* mibs/ietf/ENTITY-MIB: updated (RFC 4133).

	* mibs/ietf/BRIDGE-MIB: updated (RFC 4188).

2005-10-10  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/SSPM-MIB: added (RFC 4149).

2005-06-23  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/TUNNEL-MIB: updated (RFC 4087).

	* mibs/ietf/UDP-MIB: updated (RFC 4113).

2005-05-17  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/FC-MGMT-MIB: added (RFC 4044).

2005-05-12  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/VDSL-LINE-EXT-SCM-MIB: added (RFC 4069).

	* mibs/ietf/VDSL-LINE-EXT-MCM-MIB: added (RFC 4070).

2005-05-04  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/DOCS-IETF-SUBMGT-MIB: added (RFC 4036).

2005-04-01  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/NAT-MIB: added (RFC 4008).

	* mibs/ietf/POLICY-BASED-MANAGEMENT-MIB: added (RFC 4011).

	* mibs/ietf/TCP-MIB: updated (RFC 4022).

2005-03-04  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/INET-ADDRESS-MIB: updated (RFC 4001).

2005-01-17  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/TE-MIB: added (RFC 3970).

2004-11-30  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/parser-smi.y, lib/data.c, lib/error.c: improved handling
	of identifiers in the "NameAndNumberForm", incl. improved
	warnings if OIDs or identifiers change. Thanks to Mike, David
	Perkins, and Randy Presuhn for clarifications.
	
2004-11-29  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/parser-smi.y, lib/data.c: fixed handling of "pending"
	nodes when modules are imported right in the middle of another
	module (i.e., in MODULE-COMPLIANCE statements).
	
2004-09-29  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/ARC-MIB: added (RFC 3878).

	* mibs/ietf/SNMP-NOTIFICATION-MIB: fixed.

	* mibs/ietf/ALARM-MIB: added (RFC 3877).

2004-09-28  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/{TRIP-MIB,TRIP-TC-MIB,SCTP-MIB,IANA-ITU-ALARM-TC-MIB,
	ITU-ALARM-MIB,ITU-ALARM-TC-MIB,SNMP-NOTIFICATION-MIB,DS1-MIB,
	DS3-MIB}: added/updated (RFCs 3872,3874,3877,3895,3896).

2004-08-18  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* libsmi-0.4.3.tar.gz: released version 0.4.3. Time for a new
	release to get it out with the next stable Debian release (3.1).
	The are numerous bug fixes, improvements to smidump drivers,
	MIB updates, added MIBs, and new smilint checks.

2004-07-27  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/check.c, lib/error.c: Added checks for object groups
	containing notifications and vice versa. Thanks to Mike.

2004-07-22  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* lib/parser-smi.y, lib/error.c, lib/data.c: Added checks for
	revision clauses that reveal revisions after LAST-UPDATED,
	revisions that are not in reverse chronological order and
	missing revisions for the LAST-UPDATED timestamp. Besides that,
	revisions are now implcitly sorted in reverse chronological
	order.

2004-06-22  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/{MPLS-TE-STD-MIB,MPLS-LSR-STD-MIB,MPLS-FTN-STD-MIB,
	MPLS-LDP-STD-MIB,MPLS-LDP-ATM-STD-MIB,MPLS-LDP-FRAME-RELAY-STD-MIB,
	MPLS-LDP-GENERIC-STD-MIB}: added (RFCs 3812,3813,3814,3815).

2004-06-17  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/{Printer-MIB,Finisher-MIB,MPLS-TC-STD-MIB},
	mibs/iana/{IANA-PRINTER-MIB,IANA-FINISHER-MIB,IANA-CHARSET-MIB}:
	updated and added (RFCs 3805,3806,3808,3811).

2004-06-10  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/{ROHC-MIB,ROHC-RTP-MIB,ROHC-UNCOMPRESSED-MIB}:
	added (RFC 3816).

2004-06-02  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* tools/dump-netsnmp.c: Cleanup session in case of errors,
	Thanks to Michael Hocke.

2004-04-14  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/DIFFSERV-CONFIG-MIB: added (RFC 3747).

2004-03-03  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/APM-MIB: added (RFC 3729).

2004-02-27  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/VDSL-LINE-MIB: added (RFC 3728).


tools/smixlate.c
/*
 * smixlate.c --
 *
 *      Translate OIDs located in the input stream.
 *
 * Copyright (c) 2006 Juergen Schoenwaelder, International University Bremen.
 *
 * See the file "COPYING" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * @(#) $Id: smilint.c 1867 2004-10-06 13:45:31Z strauss $
 */

#include <config.h>

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_WIN_H
#include "win.h"
#endif

#include "smi.h"
#include "shhopt.h"

static int flags;


static char buffer[256];
static int buffer_index = 0;

static inline void buffer_add(char c)
{
    if (buffer_index >= sizeof(buffer)-2) {
	/* we should perhaps return an error and let the app flush
	   the buffer */
	exit(1);
    }
    buffer[buffer_index] = c;
    buffer_index++;
}


static inline void buffer_zap()
{
    buffer_index = 0;
}


static inline void buffer_flush()
{
    if (buffer_index) {
	buffer_add(0);
	fputs(buffer, stdout);
    }
    buffer_zap();
}


static void translate()
{
    SmiNode *smiNode;

    buffer_add(0);
    smiNode = smiGetNode(NULL, buffer);
    if (smiNode) {
	fputs(smiNode->name, stdout);
	buffer_zap();
    } else {
	buffer_flush();
    }
}


static void process(FILE *stream)
{
    int c;
    enum { TXT, NUM, NUMDOT, NUMDOTNUM, OID, OIDDOT } state = TXT;

    /*
     * Shall we require iswhite() or ispunct() before and after the
     * OID?
     *
     * TODO: - preserve and print the suffix
     *       - optionally eat/generate white space to preserve layouts
     *       - translate instance identifier to something meaningful
     *         (e.g. foobar["name",32]) where possible
     *       - provide a reverse translation service (-x) (but this is
     *         more complex since it is unclear how to identify names
     *       - parse the number into an SmiSubid vector
     *       - make sure the first sub-identifier of the OID is in the
     *         range [0:2] to reduce false positives.
     */

    while ((c = fgetc(stream)) != EOF) {
	switch (state) {
	case TXT:
	    buffer_flush();
	    if (isdigit(c)) {
		buffer_add(c);
		state = NUM;
	    } else {
		fputc(c, stdout);
		fflush(stdout);
	    }
	    break;
	case NUM:
	    if (isdigit(c)) {
		buffer_add(c);
	    } else if (c == '.') {
		buffer_add(c);
		state = NUMDOT;
	    } else {
		buffer_add(c);
		state = TXT;
	    }
	    break;
	case NUMDOT:
	    if (isdigit(c)) {
		buffer_add(c);
		state = NUMDOTNUM;
	    } else {
		buffer_add(c);
		state = TXT;
	    }
	    break;
	case NUMDOTNUM:
	    if (isdigit(c)) {
		buffer_add(c);
	    } if (c == '.') {
		buffer_add(c);
		state = OID;
	    } else {
		buffer_add(c);
		state = TXT;
	    }
	    break;
	case OID:
	    if (isdigit(c)) {
		buffer_add(c);
	    } else if (c == '.') {
		buffer_add(c);
		state = OIDDOT;
	    } else {
		translate();
		fputc(c, stdout);
		state = TXT;
	    }
	    break;
	case OIDDOT:
	    if (isdigit(c)) {
		buffer_add(c);
		state = OID;
	    } else {
		translate();
		fputc(c, stdout);
		state = TXT;
	    }
	    break;
	}
    }
}



static void usage()
{
    fprintf(stderr,
	    "Usage: smilint [options] [module or path ...]\n"
	    "  -V, --version         show version and license information\n"
	    "  -h, --help            show usage information\n"
	    "  -c, --config=file     load a specific configuration file\n"
	    "  -p, --preload=module  preload <module>\n"
	    "  -r, --recursive       print errors also for imported modules\n"
	    "  -l, --level=level     set maximum level of errors and warnings\n"
	    "  -i, --ignore=prefix   ignore errors matching prefix pattern\n"
	    "  -I, --noignore=prefix do not ignore errors matching prefix pattern\n");
}



static void help() { usage(); exit(0); }
static void version() { printf("smilint " SMI_VERSION_STRING "\n"); exit(0); }
static void config(char *filename) { smiReadConfig(filename, "smilint"); }
static void preload(char *module) { smiLoadModule(module); }
static void recursive() { flags |= SMI_FLAG_RECURSIVE; smiSetFlags(flags); }
static void level(int lev) { smiSetErrorLevel(lev); }
static void ignore(char *ign) { smiSetSeverity(ign, 128); }
static void noignore(char *ign) { smiSetSeverity(ign, -1); }



int main(int argc, char *argv[])
{
    int i;

    static optStruct opt[] = {
	/* short long              type        var/func       special       */
	{ 'h', "help",           OPT_FLAG,   help,          OPT_CALLFUNC },
	{ 'V', "version",        OPT_FLAG,   version,       OPT_CALLFUNC },
	{ 'c', "config",         OPT_STRING, config,        OPT_CALLFUNC },
	{ 'p', "preload",        OPT_STRING, preload,       OPT_CALLFUNC },
	{ 'r', "recursive",      OPT_FLAG,   recursive,     OPT_CALLFUNC },
	{ 'l', "level",          OPT_INT,    level,         OPT_CALLFUNC },
	{ 'i', "ignore",         OPT_STRING, ignore,        OPT_CALLFUNC },
	{ 'I', "noignore",       OPT_STRING, noignore,      OPT_CALLFUNC },
	{ 0, 0, OPT_END, 0, 0 }  /* no more options */
    };
    
    for (i = 1; i < argc; i++)
	if ((strstr(argv[i], "-c") == argv[i]) ||
	    (strstr(argv[i], "--config") == argv[i])) break;
    if (i == argc) 
	smiInit("smixlate");
    else
	smiInit(NULL);

    flags = smiGetFlags();
    flags |= SMI_FLAG_ERRORS;
    flags |= SMI_FLAG_NODESCR;
    smiSetFlags(flags);

    optParseOptions(&argc, argv, opt, 0);

    for (i = 1; i < argc; i++) {
	if (smiLoadModule(argv[i]) == NULL) {
	    fprintf(stderr, "smilint: cannot locate module `%s'\n",
		    argv[i]);
	    smiExit();
	    exit(1);
	}
    }


    process(stdin);

    smiExit();

    return 0;
}
